### PR TITLE
feat(sink): enable user-defined primary key for upsert sink

### PR DESCRIPTION
**Pull request by xx01cyx**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Add a `primary_key` field in with options when creating a sink. The primary key columns should be delimited by `,`. Here is an example:  `CREATE SINK si AS SELECT v1, v2, v3 FROM t WITH (primary_key = 'v1, v2', ......);`
- Rename `format = 'append_only'` to `type = 'append-only'` when specifying the append-only attribute of a sink.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [ ] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Connector (sources & sinks)
- SQL commands, functions, and operators

##### Sink primary key

The user can now create a sink with primary keys as follows:

```SQL
CREATE TABLE t (v1 int, v2 int, v3 int);
CREATE SINK si FROM t WITH (
    connector = 'xxx', 
    primary_key = 'v1, v2',
	......
);
```

##### Sink type

The following SQL is no longer valid:

```SQL
CREATE SINK si FROM t WITH (
    format = 'append_only', 
	......
);
```

Instead, the user should write:

```SQL
CREATE SINK si FROM t WITH (
    type = 'append-only', 
	......
);
```

</details>

### CODE REVIEW

File: src/connector/src/sink/kafka.rs [hunk 1]
```diff
@@ -134,6 +140,20 @@ impl<const APPEND_ONLY: bool> KafkaSink<APPEND_ONLY> {
...
+    pub async fn validate(config: KafkaConfig, pk_indices: Vec<usize>) -> Result<()> {
+        // For upsert Kafka sink, the primary key must be defined.
+        if !APPEND_ONLY && pk_indices.is_empty() {
+            return Err(SinkError::Config(anyhow!(
+                "primary key not defined for upsert kafka sink (please define in `primary_key` field)"
+            )));
+        }
+
+        // Try Kafka connection.
+        KafkaTransactionConductor::new(config).await?;
```
Comment (tabVersion) [inline src/connector/src/sink/kafka.rs:152]: can we reuse the conductor instance? it has some overhead to recreate a connection to kafka broker

File: src/frontend/src/optimizer/plan_node/stream_sink.rs [hunk 2]
```diff
@@ -112,35 +113,50 @@ impl StreamSink {
...
-        let (pk, stream_key) = derive_pk(input, user_order_by, &columns);
-
-        if sink_type == SinkType::Upsert && pk.is_empty() {
-            return Err(ErrorCode::SinkError(Box::new(Error::new(
-                ErrorKind::InvalidInput,
-                "No primary key for the upsert sink. Please include the primary key explicitly in sink definition or make the sink append-only.",
-            )))
-            .into());
-        }
+        let (pk, _) = derive_pk(input, user_order_by, &columns);
```
Comment (tabVersion) [inline src/frontend/src/optimizer/plan_node/stream_sink.rs:120]: IIRC, no need to derive pk now? maybe mark the field deprecated.

**Review by tabVersion — COMMENTED** [review]
sink part LGTM
waiting for comments about the syntax

File: src/frontend/src/optimizer/plan_node/stream_sink.rs (hunk 2 above)
Comment (xx01cyx) [reply src/frontend/src/optimizer/plan_node/stream_sink.rs:120]: cc. @st1page

**Comment by tabVersion** [comment]
also have a concern

```
create sink s as 
  select a.id, b.id from a join b on a.sth = b.sth
with (
  primary_key = 'a.id, b.id',
)
```

is it acceptable? what is the pk to the downstream?

File: src/frontend/src/optimizer/plan_node/stream_sink.rs (hunk 2 above)
Comment (st1page) [reply src/frontend/src/optimizer/plan_node/stream_sink.rs:120]: I think so. But I prefer to store "down_stream_pk" and "plan_original_pk" in the SinkCatalog proto at least for debugging

File: proto/catalog.proto [hunk 3]
```diff
@@ -76,8 +76,8 @@ message Sink {
   repeated common.ColumnOrder pk = 6;
```
Comment (st1page) [inline proto/catalog.proto:76]:
```suggestion
    // primary key indices derived from the SQL.
    repeated common.ColumnOrder plan_pk = 6;
```

File: src/connector/src/sink/kafka.rs (hunk 1 above)
Comment (xx01cyx) [reply src/connector/src/sink/kafka.rs:152]: May refine this in the future.

**Comment by xx01cyx** [comment]
> also have a concern
> 
> ```
> create sink s as 
>   select a.id, b.id from a join b on a.sth = b.sth
> with (
>   primary_key = 'a.id, b.id',
> )
> ```
> 
> is it acceptable? what is the pk to the downstream?

We may write the SQL like this:

```SQL
create sink s as 
  select a.id as aid, b.id as bid from a join b on a.sth = b.sth
with (primary_key = 'aid, bid');
```

and the primary key will be `aid` and `bid`.

**Comment by wenym1** [comment]
> create sink s as 
  select a.id, b.id from a join b on a.sth = b.sth
with (
  primary_key = 'a.id, b.id',
)

I think the SQL should report an error like `ambiguous column name`? Because in the downstream sink there are two columns with the same name `id`.

**Comment by xx01cyx** [comment]
> > create sink s as
> > select a.id, b.id from a join b on a.sth = b.sth
> > with (
> > primary_key = 'a.id, b.id',
> > )
> 
> I think the SQL should report an error like `ambiguous column name`? Because in the downstream sink there are two columns with the same name `id`.

RW will actually prevent this. We must specify an alias for `a.id` and `b.id`, or the DDL will fail.

**Review by tabVersion — APPROVED** [review]

File: proto/stream_plan.proto [hunk 4]
```diff
@@ -142,8 +142,8 @@ message SinkDesc {
   string name = 2;
   string definition = 3;
   repeated plan_common.ColumnDesc columns = 4;
-  repeated common.ColumnOrder pk = 5;
-  repeated uint32 stream_key = 6;
+  repeated common.ColumnOrder plan_pk = 5;
```
Comment (wenym1) [inline proto/stream_plan.proto:145]: Will we use the plan_pk anywhere?

Comment (wenym1) [reply proto/stream_plan.proto:145]: And do we enforce any constraint on the relation between `plan_pk` and `downstream_pk` in this PR?

**Review by wenym1 — APPROVED** [review]
Generally LGTM. 

May need to search all usage of `sink.mode` and replace it with `type`. As far as I know, in the `integration_test/iceberg-sink` there is a legacy create iceberg sink SQL that still uses the old configuration format.

Besides, we'd better have a tracking issue in any related repos such as kube-bench and risingwave-demo that may contain the create sink SQL with old configuration format.

File: proto/stream_plan.proto (hunk 4 above)
Comment (xx01cyx) [reply proto/stream_plan.proto:145]: > And do we enforce any constraint on the relation between `plan_pk` and `downstream_pk` in this PR?

No. `downstream_pk` is specified by the user, and the user can specify anything he or she wants, as long as the columns are valid.

File: e2e_test/sink/iceberg_sink.slt [hunk 5]
```diff
@@ -7,7 +7,8 @@ CREATE MATERIALIZED VIEW mv6 AS SELECT * FROM t6;
 statement ok
 CREATE SINK s6 AS select mv6.v1 as v1, mv6.v2 as v2, mv6.v3 as v3 from mv6 WITH (
     connector = 'iceberg',
-    sink.mode='append-only',
+    type = 'upsert',
+    primary_key = 'v1',
```
[the remark this answers is not in the document]
Comment (xx01cyx) [reply e2e_test/sink/iceberg_sink.slt:11]: IMO `a.b` implies that `b` is a subcategory (or sub-configuration) under `a`, so I prefer `primary_key`. Remember we also have a  `force_append_only` option. cc. @tabVersion @st1page @fuyufjh

**Pull request merged**